spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DatastoreServiceObjectToKeyFactory.java

```diff
@@ -76,9 +77,18 @@ public Key getKeyFromObject(Object entity,
 					"Cannot construct key for entity types without ID properties: "
 							+ entity.getClass());
 		}
-		return getKeyFromId(
-				datastorePersistentEntity.getPropertyAccessor(entity).getProperty(idProp),
-				datastorePersistentEntity.kindName());
+		PersistentPropertyAccessor accessor = datastorePersistentEntity
+				.getPropertyAccessor(entity);
+		Object idVal = accessor.getProperty(idProp);
+		if (idVal == null) {
+			Key key = this.datastore.allocateId(getKeyFactory()
```

Comment:
Since you're allocating IDs here, the method should be called something like `getOrCreateKeyFromObject`.

Reply:
@meltsufin changes made, PTAL.

Reply:
Thinking about it a bit more, I don't really like a single method having the responsibility for both extracting and allocating keys. The decision whether/how to allocate the keys should probably be at the caller level.
Perhaps, change this method back to `getKeyFromObject` and just return null, if null; and also just add a method called `allocateKeyForObject`. The caller would decide whether it needs to use `allocateKey`. WDYT?

Reply:
Yes, it sounds good to me to let the caller decide. I'll make a follow up issue for adding the mechanism by which the user chooses (like an annotation `@AutoAllocateId` )

Reply:
#891